### PR TITLE
Non-repudiation: add persistent, SQL-backed signed commands repo

### PR DESCRIPTION
**Pull request by stefanobaghino-da**
Closes https://github.com/digital-asset/daml/issues/8634

Contributes to https://github.com/digital-asset/daml/issues/8658 by adding a persistent store for keys as well (but does not add an interface to upload new ones).

changelog_begin
changelog_end

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: runtime-components/non-repudiation-postgresql/src/test/scala/com/daml/nonrepudiation/postgresql/NonRepudiationProxyConformance.scala [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+import sun.security.tools.keytool.CertAndKeyGen
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration.DurationInt
+
+final class NonRepudiationProxyConformance
+    extends AsyncFlatSpec
+    with Matchers
+    with EitherValues
+    with PostgresAroundAll {
```
Comment (cocreature) [inline runtime-components/non-repudiation-postgresql/src/test/scala/com/daml/nonrepudiation/postgresql/NonRepudiationProxyConformance.scala:45]: Should we have an in-memory backend (e.g. h2) for testing (by users) and debugging?

**Review by cocreature — APPROVED** [review]
Nice work, thank you!

File: runtime-components/non-repudiation-postgresql/src/main/scala/com/daml/nonrepudiation/postgresql/package.scala [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    Get[Array[Byte]].map { bytes =>
+      val keySpec = new X509EncodedKeySpec(bytes)
+      val keyFactory = KeyFactory.getInstance("RSA")
+      keyFactory.generatePublic(keySpec)
+    }
+
+  implicit val putPublicKey: Put[PublicKey] =
+    Put[Array[Byte]].contramap(_.getEncoded)
+
+  def slf4jLogHandler(logger: Logger): LogHandler =
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-postgresql/src/main/scala/com/daml/nonrepudiation/postgresql/package.scala:45]: In the triggers we use the built-in log handler, which according to a comment

```
/**
 * A LogHandler that writes a default format to a JDK Logger. This is provided for demonstration
 * purposes and is not intended for production use.
 * @group Constructors
 */
```

So one of the things I wanted to do was to extract this piece and use it in both places.

File: runtime-components/non-repudiation-postgresql/src/main/scala/com/daml/nonrepudiation/postgresql/PostgresqlKeyRepository.scala [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    logHandler: LogHandler
+) extends KeyRepository {
+
+  private val table = Fragment.const(s"${Tables.Prefix}_keys")
+
+  private def getKey(fingerprint: FingerprintBytes): Fragment =
+    fr"""select "key" from """ ++ table ++ fr"where fingerprint = $fingerprint"
+
+  override def get(fingerprint: FingerprintBytes): Option[PublicKey] =
+    getKey(fingerprint).query[PublicKey].option.transact(transactor).unsafeRunSync()
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-postgresql/src/main/scala/com/daml/nonrepudiation/postgresql/PostgresqlKeyRepository.scala:25]: I think it makes sense to bubble up the side-effecting part to the intereceptor, but we can't go higher than that. 😉

File: runtime-components/non-repudiation-postgresql/src/main/scala/com/daml/nonrepudiation/postgresql/Tables.scala [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import com.daml.nonrepudiation.{CommandIdString, KeyRepository, SignedPayloadRepository}
+import doobie.hikari.HikariTransactor
+import doobie.util.log.LogHandler
+import org.flywaydb.core.Flyway
+
+object Tables {
+
+  val Prefix = "nonrepudiation"
+
+  def initialize(transactor: HikariTransactor[IO])(implicit logHandler: LogHandler): Tables =
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-postgresql/src/main/scala/com/daml/nonrepudiation/postgresql/Tables.scala:18]: Tracked by https://github.com/digital-asset/daml/issues/8728

File: runtime-components/non-repudiation-postgresql/src/main/scala/com/daml/nonrepudiation/postgresql/PostgresqlKeyRepository.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (cocreature) [reply runtime-components/non-repudiation-postgresql/src/main/scala/com/daml/nonrepudiation/postgresql/PostgresqlKeyRepository.scala:25]: I’m more worried about the `Sync` part than the `unsafe` here. We use `unsafeRunAsync` fairly liberally.

File: runtime-components/non-repudiation-postgresql/src/test/scala/com/daml/nonrepudiation/postgresql/NonRepudiationProxyConformance.scala (hunk 1 above)
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-postgresql/src/test/scala/com/daml/nonrepudiation/postgresql/NonRepudiationProxyConformance.scala:45]: Should we simply expose the map-backed implementation of the repositories or did you have something else in mind?

Comment (cocreature) [reply runtime-components/non-repudiation-postgresql/src/test/scala/com/daml/nonrepudiation/postgresql/NonRepudiationProxyConformance.scala:45]: map-backed seems perfectly fine. Not sure if it’s easier to keep that working or to have an h2 implementation. happy to leave it up to you.

File: runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sha256 [hunk 5]
```diff
@@ -0,0 +1 @@
+33290da121c6e8e6a7f08a2d177adde7f2e1dabe587806b7f99b5b24ab5eb85f
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sha256:1]: Uh, you're right, I thought it was done transparently by FlyWay. Thanks!

File: runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    algorithm text not null,
+    fingerprint bytea not null,
+    payload bytea not null,
+    signature bytea not null
+);
+
+create index ${tables.prefix}_signed_payloads_command_id_index on ${tables.prefix}_signed_payloads using hash(command_id);
+
+create table ${tables.prefix}_keys(
+    fingerprint bytea not null,
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql:15]: It makes a lot of sense of course. The issue is that hash indices do not support uniqueness (at least in Postgresql), so I favored that over having a primary index. I can probably leave the normal B-tree index and see later if there is any advantage in using something different.

File: runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql [hunk 7]
```diff
@@ -0,0 +1,19 @@
+-- Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+create table ${tables.prefix}_signed_payloads(
+    command_id text,
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql:5]: Absolutely, good catch, thanks!

File: runtime-components/non-repudiation-postgresql/src/main/scala/com/daml/nonrepudiation/postgresql/package.scala [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+  implicit val getCommandIdString: Get[CommandIdString] =
+    Get[String].map(CommandIdString.wrap)
+
+  implicit val putCommandIdString: Put[CommandIdString] =
+    Put[String].contramap(identity)
+
+  implicit val getPublicKey: Get[PublicKey] =
+    Get[Array[Byte]].map { bytes =>
+      val keySpec = new X509EncodedKeySpec(bytes)
+      val keyFactory = KeyFactory.getInstance("RSA")
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-postgresql/src/main/scala/com/daml/nonrepudiation/postgresql/package.scala:38]: I implemented this part rather quickly, I'll have a second look.

File: runtime-components/non-repudiation-postgresql/src/test/scala/com/daml/nonrepudiation/postgresql/NonRepudiationProxyConformance.scala (hunk 1 above)
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-postgresql/src/test/scala/com/daml/nonrepudiation/postgresql/NonRepudiationProxyConformance.scala:45]: I would keep the map-backed implementation where it is right now, it should be fine for testing without adding development overhead. We can easily revise this decision in the future. But thanks for raising this!

File: runtime-components/non-repudiation-postgresql/src/main/scala/com/daml/nonrepudiation/postgresql/package.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-postgresql/src/main/scala/com/daml/nonrepudiation/postgresql/package.scala:45]: I spun this off in https://github.com/digital-asset/daml/pull/8731

[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-postgresql/src/main/scala/com/daml/nonrepudiation/postgresql/package.scala]: Addressed by https://github.com/digital-asset/daml/pull/8723/commits/b01d8fb0ef3f1bc00358d89988ecd002e617ed06 after merging https://github.com/digital-asset/daml/pull/8731 and rebasing on the latest `main`.

File: runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql (hunk 6 above)
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql:15]: I guess we can also make this column fixed in length.

File: runtime-components/non-repudiation-postgresql/src/main/scala/com/daml/nonrepudiation/postgresql/package.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-postgresql/src/main/scala/com/daml/nonrepudiation/postgresql/package.scala:36]: I supposed we should add the key algorithm to the key repository, alongside the key itself, in order to support more than an hard-coded one, right?

[the remark this answers is not in the document]
Comment (cocreature) [reply runtime-components/non-repudiation-postgresql/src/main/scala/com/daml/nonrepudiation/postgresql/package.scala:36]: sounds sensible, or infer it from the cert. No idea if that’s possible for x509.

File: runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql (hunk 7 above)
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql:5]: Good point. Command identifiers are validated by the Ledger API and are `LedgerString`s, which at the current status means that their size is limited to 255 characters. I think it's safe to set that here as well. Thanks!

File: runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    algorithm text not null,
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql:6]: You're right. I had a look [here](https://docs.oracle.com/javase/9/docs/specs/security/standard-names.html) and the longer name I can find is a bit more than 40 characters. Would 64 be an acceptable size?

File: runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql (hunk 7 above)
[the remark this answers is not in the document]
Comment (cocreature) [reply runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql:5]: are you sure? my understanding was that party ids are limited to 255 but LedgerStrings in general don’t have a length limit.

**Comment by stefanobaghino-da** [comment]
Converting to draft as I tackle existing comments and review the existing interfaces a bit.

File: runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql (hunk 9 above)
[the remark this answers is not in the document]
Comment (cocreature) [reply runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql:6]: 64 seems sensible to me :+1:

File: runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql (hunk 7 above)
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql:5]: Spun off in https://github.com/digital-asset/daml/pull/8736

I'll rebase on that and add a field for the signature verification timestamp.

[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql:5]: I referred to this

https://github.com/digital-asset/daml/blob/5a96ad0c9348a81dd5b4256834c7753adf190554/daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/IdString.scala#L312-L319

and this

https://github.com/digital-asset/daml/blob/5a96ad0c9348a81dd5b4256834c7753adf190554/daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/IdString.scala#L199-L203

The limit seems to be 255, right?

File: runtime-components/non-repudiation-postgresql/src/main/scala/com/daml/nonrepudiation/postgresql/package.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-postgresql/src/main/scala/com/daml/nonrepudiation/postgresql/package.scala:36]: I spun off switching from bare public keys to certificates in https://github.com/digital-asset/daml/pull/8739.

Once that is merged (note: it builds on top of https://github.com/digital-asset/daml/pull/8736 to minimize merge conflicts) I'll rebase on the latest `main` and address this.

🙇🏻

**Comment by stefanobaghino-da** [comment]
Closed by mistake due to a wrong rebase and push.

File: runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql (hunk 7 above)
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql:5]: Addressed by https://github.com/digital-asset/daml/pull/8723#discussion_r569223780 after rebasing

File: runtime-components/non-repudiation-postgresql/src/main/scala/com/daml/nonrepudiation/postgresql/package.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-postgresql/src/main/scala/com/daml/nonrepudiation/postgresql/package.scala]: Addressed by https://github.com/digital-asset/daml/pull/8723/commits/b2993b4b92b6751b73cfc59b415ef97edb5fa243 after rebasing

File: runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql (hunk 7 above)
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql:5]: If there are no objections I'll limit this to 255 characters. 🙂

[the remark this answers is not in the document]
Comment (cocreature) [reply runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql:5]: :+1:

File: runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql (hunk 6 above)
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql:16]: The support for `bit(n)` for `n > 1` is a bit (sic) of an [issue](https://github.com/pgjdbc/pgjdbc/issues/908) apparently right now for Postgresql, I'll defer fixing the size of this for now.

File: runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql (hunk 7 above)
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql]: Addressed by https://github.com/digital-asset/daml/pull/8723/commits/19c5a1d5fe2780ac71e283b9bed6648948823eb9

File: runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql (hunk 9 above)
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sql]: Addressed by https://github.com/digital-asset/daml/pull/8723/commits/19c5a1d5fe2780ac71e283b9bed6648948823eb9

**Review by cocreature — APPROVED** [review]
Thanks!

File: runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sha256 (hunk 5 above)
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sha256]: I deleted this for now, I'll defer introducing the relevant checks as part of #8728.

[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-postgresql/src/main/resources/com/daml/nonrepudiation/postgresql/V1__init.sha256]: Addressed by https://github.com/digital-asset/daml/pull/8723/commits/e312e7af9e2894a49b3e4598f065cc01535ca54f

**Pull request merged**